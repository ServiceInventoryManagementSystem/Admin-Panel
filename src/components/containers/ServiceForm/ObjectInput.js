import React, { Component } from 'react';
import _s from 'assets/css/ServiceForm.css';
import TextField from 'material-ui/TextField';
import DatePicker from 'material-ui/DatePicker';
import Toggle from 'material-ui/Toggle';
import { mapAndConnect, IManagedService, ManagedService} from "services";
import {Redirect} from "react-router";

import DropDownMenu from 'material-ui/DropDownMenu';
import MenuItem from 'material-ui/MenuItem';


import { get, set, has, cloneDeep } from 'lodash';


export class ObjectInput extends Component{
  constructor(props){
    super(props);

    this.state = {
      value: {}
    }
  }

  componentDidMount(){
    this._updateFields(this.props);
  }

  componentWillReceiveProps(nextProps){
    this._updateFields(nextProps);
  }

  _updateFields(props){
    let nmap = {};
    React.Children.forEach(props.children, child => {
      let props = child.props;
        
      if(props.name){
        if(!has(this.state.value, props.name)){
          set(nmap, props.name, props.value != null ? props.value : props.defaultValue);
        }else{
          set(nmap, props.name, get(this.state.value, props.name));
        }
      }
    });
    if(props.object != null){
      
    }
    this.setState({
      value: nmap
    })
  }

  onFieldChange(name, nval){
    set(this.state.value, name, nval);



    let object = cloneDeep(this.state.value);
    this.setState({
      value: object
    });
    
    if(this.props.map) {
      object = this.props.map(object);
    }
    
    if(this.props.onChange){
      this.props.onChange(object);
    }
  }


  render(){
    const inputs = React.Children.map(this.props.children, (field) => {
      let onChange = (...a) => {
        this.onFieldChange(field.props.name, a[1]);
        if(field.props.onChange){
          field.props.onChange(...a);
        }
      }

      return React.cloneElement(field, {
        onChange: onChange
      });
    });
    return (
      <div>
        {inputs}
      </div>
    );
  }
}