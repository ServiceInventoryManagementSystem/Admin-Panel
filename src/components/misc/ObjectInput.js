import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Toggle from 'material-ui/Toggle';
import { mapAndConnect, IManagedService, ManagedService} from "services";
import {Redirect} from "react-router";

import DropDownMenu from 'material-ui/DropDownMenu';
import MenuItem from 'material-ui/MenuItem';


import { get, set, has, cloneDeep } from 'lodash';
import { paths } from 'common/utils';


export class ObjectInput extends Component{
  constructor(props){
    super(props);

    this.state = {
      value: {}
    }
  }

  componentDidMount(){
    this._updateFields(this.props);
  }

  componentWillReceiveProps(nextProps){
    this._updateFields(nextProps);
  }

  isControlled(){
    return this.props.value != null;
  }

  _updateFields(props){
    let nmap = {};
    if(props.value != null){
      for(let atr of paths(props.value) ){
        set(nmap, atr, get(props.value, atr));
      }
    }
    this.setState({
      value: nmap
    })
  }

  onFieldChange(name, nval){
    set(this.state.value, name, nval);
    let object = cloneDeep(this.state.value);
    this.setState({
      value: object
    });
    
    if(this.props.map) {
      object = this.props.map(object);
    }

    if(this.props.onChange){
      this.props.onChange(object);
    }
  }


  render(){

    const inputs = React.Children.map(this.props.children, (field, idx) => {
      if(field.props.name){
        let onChange = (...a) => {
          let value = a[1];
          if(field.props.mapValue){
            value = field.props.mapValue(...a);
          }else if(this.props.mapValues){
            value = this.props.mapValues(...a);
          }

          this.onFieldChange(field.props.name, value);
          if(field.props.onChange){
            field.props.onChange(...a);
          }
          
          if(field.props.onCheck){
            field.props.onCheck(...a);
          }

        }
        return React.cloneElement(field, {
          onChange: onChange,
          onCheck: onChange,
          value: this.isControlled() ? get(this.state.value, field.props.name) : undefined,
          checked: this.isControlled() ? get(this.state.value, field.props.name) : undefined,
          key: field.key || field.props.key || idx
        });
      }

      return field;
    });
    return (
      <div style={{display: "inline-block"}}>
        {inputs}
      </div>
    );
  }
}


ObjectInput.defaultProps = {
  mapValues: (event, v) => {
    if(v != null){return v;}
    return event.target.value || event.target.checked;
  }
}

ObjectInput.propTypes = {
  value: PropTypes.object,
  mapValues: PropTypes.func
}

