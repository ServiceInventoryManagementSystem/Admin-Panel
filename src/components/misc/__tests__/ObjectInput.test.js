import React from 'react';
import ReactDOM from 'react-dom';
import { ObjectInput } from 'components/misc/ObjectInput.js';

import { mount, shallow } from 'enzyme';



describe('ObjectInput', () => {
  
  it('renders with no props', () => {
    const wrapper = shallow(<ObjectInput />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders with inputs', () => {
    const wrapper = shallow(
      <ObjectInput>
        <input name="name" type="text" />
        <input name="href" type="text" />
      </ObjectInput>
    );
    expect(wrapper).toMatchSnapshot();
  });

});
