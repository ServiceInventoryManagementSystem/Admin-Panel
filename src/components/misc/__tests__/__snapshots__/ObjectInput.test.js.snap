// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectInput renders with inputs 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
>
    <input
        name="name"
        type="text"
    />
    <input
        name="href"
        type="text"
    />
</ObjectInput>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          checked={undefined}
          name="name"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value={undefined}
/>,
        <input
          checked={undefined}
          name="href"
          onChange={[Function]}
          onCheck={[Function]}
          type="text"
          value={undefined}
/>,
      ],
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0/.0",
        "nodeType": "host",
        "props": Object {
          "checked": undefined,
          "name": "name",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": "1/.1",
        "nodeType": "host",
        "props": Object {
          "checked": undefined,
          "name": "href",
          "onChange": [Function],
          "onCheck": [Function],
          "type": "text",
          "value": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            checked={undefined}
            name="name"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value={undefined}
/>,
          <input
            checked={undefined}
            name="href"
            onChange={[Function]}
            onCheck={[Function]}
            type="text"
            value={undefined}
/>,
        ],
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0/.0",
          "nodeType": "host",
          "props": Object {
            "checked": undefined,
            "name": "name",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": "1/.1",
          "nodeType": "host",
          "props": Object {
            "checked": undefined,
            "name": "href",
            "onChange": [Function],
            "onCheck": [Function],
            "type": "text",
            "value": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`ObjectInput renders with no props 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ObjectInput
    mapValues={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": undefined,
      "style": Object {
        "display": "inline-block",
      },
    },
    "ref": null,
    "rendered": null,
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "style": Object {
          "display": "inline-block",
        },
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
